sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala

```diff
@@ -147,10 +163,15 @@ object PartitionPruning extends Rule[LogicalPlan] with PredicateHelper {
       case _ => fallbackRatio
     }
 
+    val estimatePruningSideSize = filterRatio * partPlan.stats.sizeInBytes.toFloat
     // the pruning overhead is the total size in bytes of all scan relations
     val overhead = otherPlan.collectLeaves().map(_.stats.sizeInBytes).sum.toFloat
-
-    filterRatio * partPlan.stats.sizeInBytes.toFloat > overhead.toFloat
+    if (canBuildBroadcast) {
+      estimatePruningSideSize > overhead
+    } else {
+      // We need to ensure that the otherPlan can collect to the driver.
+      canBroadcastBySize(otherPlan, SQLConf.get) && estimatePruningSideSize * 0.04 > overhead
```

Comment:
what is this `0.04`?

Reply:
This is to ensure that the pruning side is much lager, after all, we cannot reuse broadcast. For example.

If overhead = 10M

  | canBuildBroadcast | canNotBuildBroadcast
-- | -- | --
hasBenefit=true | estimatePruningSideSize > 20M | estimatePruningSideSize > 500M